cmusatyalab/openface
# Convert the OpenFace model from Lua torch to Pytorch

### What does this PR do?
This PR rewrites the OpenFace model in PyTorch and provides scripts to convert the trained model weights `nn4.small2.v1.t7` to a PyTorch state_dict. This PR also provides examples of comparing images and training classifiers using the new model.

#### Summary of Changes:
- `openface/openfacenet.py`: Add PyTorch model definition
- `openface/align_dlib.py`: Support dlib's CNN face detector; Make upsampling count a tunable parameter
- `models/get-models.sh`: Download dlib's mmod_human_face_detector model and the converted PyTorch OpenFace model weights
- `batch-represent/batch_represent.py`: Generate representations for an image dataset and store the data and labels in .csv files
- `demos/compare_new.py`, `demos/classifier_new.py`: Add new examples of using the PyTorch model
- `conversion/test_luatorch.lua`, `conversion/convert_to_pytorch.py`: Add scripts to convert the trained model weights to a PyTorch state_dict (not necessary for using the PyTorch model)

### Where should the reviewer start?
- Build a Docker image from `Dockerfile` under the project root directory
```
sudo docker build . -t newface
```
- Run a Docker container with GPU enabled
```
sudo docker run --rm -it --gpus all newface
```

### How should this PR be tested?
- Run the comparison demo
```
python3 demos/compare_new.py images/examples/{lennon*,clapton*}
```
- Generate representations in batches for an image dataset, e.g. a raw image directory `data/mydataset/raw/`
```
python3 batch-represent/batch_represent.py -i data/mydataset/raw/ -o data/mydataset/feats/ --align_out data/mydataset/aligned/
```
- Train a new SVM classifier
```
python3 demos/classifier_new.py train data/mydataset/feats/
```
- Run the new classifier
```
python3 demos/classifier_new.py infer data/mydataset/feats/classifier.pkl data/mydataset/test/*
```
Note that only the new PyTorch model and the new examples should work. Backward compatibility with the previous version has not been carefully tested yet. The model training and the web demo are currently not supported in the update.

## Questions:

+ Do the docs need to be updated?

Yes

+ Does this PR add new (Python) dependencies?

Yes
